### PR TITLE
UCT/DC: Fix possible DCI leaks in FC

### PR DESCRIPTION
**Pull request by brminich**
Fix possible DCI leaks. When grant is received all pendings are processed. If some other eps (which are blocked by FC) are scheduled for dci allocation, they will get DCIs as well (while TX is blocked).
So ep should not be scheduled for DCI allocation if FC window is empty.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3110/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/674/ for details.

File: src/uct/ib/dc/base/dc_ep.h [hunk 1]
```diff
@@ -166,7 +166,11 @@ static inline void uct_dc_iface_dci_put_dcs(uct_dc_iface_t *iface, uint8_t dci)
      * move the group to the 'wait for dci alloc' state
      */
     ucs_arbiter_group_desched(uct_dc_iface_tx_waitq(iface), &ep->arb_group);
-    ucs_arbiter_group_schedule(uct_dc_iface_dci_waitq(iface), &ep->arb_group);
+    if (ep->fc.fc_wnd > 0) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.h]: nice catch.
does it make sense to extract this code to common function?
is it possible to reproduce this bug with a unit test?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2627/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2628/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:ornl:retest

File: src/uct/ib/dc/base/dc_ep.h (hunk 1 above)
Comment (brminich) [inline src/uct/ib/dc/base/dc_ep.h]: ok

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3122/ for details.

**Comment by brminich** [comment]
Yes, but there is a chance that grants are not arriving for several eps, which seems like a wasting of DCIs. (theoretically, in the worst case, all DCIs can be blocked by eps waiting for grants)

**Comment by alex-mikheev** [comment]
But the grants will eventually arrive right ?

**Comment by brminich** [comment]
yes

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2632/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3131/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3141/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3144/ for details.

**Comment by brminich** [comment]
Mellanox failure is unrelated (reported in #1224)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/694/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2640/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

**Comment by brminich** [comment]
@yosefe, @alex-mikheev, please check the update. MLX failure is unrelated

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2641/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3146/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3147/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3151/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3158/ for details.

**Pull request merged**